cpp/src/groupby/hash/groupby_kernels.cuh

```diff
@@ -0,0 +1,123 @@
+ * between key rows. For example, comparing two keys `k0` and `k1` will compare
+ * the two rows `input_keys[k0] ?= input_keys[k1]`
+ *
+ * Likewise, we store the row indices for the hash maps "values". These indices
+ * index into the `output_values` table. For a given key `k` (which is an index
+ * into `input_keys`), the corresponding value `v` indexes into `output_values`
+ * and stores the result of aggregating rows from `input_values` from rows of
+ * `input_keys` equivalent to the row at `k`.
+ *
+ * The exact size of the result is not known a priori, but can be upper bounded
+ * by the number of rows in `input_keys` & `input_values`. Therefore, it is
+ * assumed `output_values` has sufficient storage for an equivalent number of
+ * rows. In this way, after all rows are aggregated, `output_values` will likely
+ * be "sparse", meaning that not all rows contain the result of an aggregation.
+ *
+ * @tparam skip_rows_with_nulls Indicates if rows in `input_keys` containing
+ * null values should be skipped. It `true`, it is assumed `row_bitmask` is a
+ * bitmask where bit `i` indicates the presence of a null value in row `i`.
+ * @tparam Map The type of the hash map
+ */
+template <bool skip_rows_with_nulls, typename Map>
+struct compute_single_pass_aggs {
+  Map map;
+  size_type num_keys;
+  table_device_view input_values;
+  mutable_table_device_view output_values; 
+  aggregation::Kind const* __restrict__ aggs;
+  bitmask_type const*  __restrict__ row_bitmask;
+
+  /**
+   * @brief Construct a new compute_single_pass_aggs functor object
+   * 
+   * @param map Hash map object to insert key,value pairs into.
+   * @param num_keys The number of rows in input keys table
+   * @param input_values The table whose rows will be aggregated in the values 
+   * of the hash map
+   * @param output_values Table that stores the results of aggregating rows of
+   * `input_values`.
+   * @param aggs The set of aggregation operations to perform accross the 
+   * columns of the `input_values` rows
+   * @param row_bitmask Bitmask where bit `i` indicates the presence of a null
+   * value in row `i` of input keys. Only used if `skip_rows_with_nulls` is `true`
+   */
+  compute_single_pass_aggs(
+    Map map,
+    size_type num_keys,
+    table_device_view input_values,
+    mutable_table_device_view output_values, 
+    aggregation::Kind const* aggs,
+    bitmask_type const* row_bitmask)
+  : map(map),
+    num_keys(num_keys),
+    input_values(input_values),
+    output_values(output_values),
+    aggs(aggs),
+    row_bitmask(row_bitmask)
+  {}
+
+  __device__ void operator()(size_type i) {
+    if (not skip_rows_with_nulls or
+        cudf::bit_is_set(row_bitmask, i))
+    {
+      auto result = map.insert(thrust::make_pair(i, i));
+
+      experimental::detail::aggregate_row<true, true>(
+        output_values, result.first->second, input_values, i, aggs);
+    }
+  }
+};
+
+// TODO (dm): variance kernel
```

Comment:
In this PR or a later PR?

Reply:
Aggregations that depend on other aggregations deferred to later PR.